src/tallies/tally_scoring.cpp

```diff
@@ -0,0 +1,2347 @@
+{
+  // Save the original delayed group bin
+  auto& tally {*model::tallies[i_tally]};
+  auto i_filt = tally.filters(tally.delayedgroup_filter_);
+  auto& dg_match {simulation::filter_matches[i_filt]};
+  auto i_bin = dg_match.i_bin_;
+  auto original_bin = dg_match.bins_[i_bin];
+  dg_match.bins_[i_bin] = d_bin;
+
+  // Determine the filter scoring index
+  auto filter_index = 0;
+  for (auto i = 0; i < tally.filters().size(); ++i) {
+    auto i_filt = tally.filters(i);
+    auto& match {simulation::filter_matches[i_filt]};
+    auto i_bin = match.i_bin_;
+    filter_index += match.bins_[i_bin] * tally.strides(i);
+  }
+
+  // Update the tally result
+  #pragma omp atomic
+  tally.results_(filter_index, score_index, RESULT_VALUE) += score;
+
+  // Reset the original delayed group bin
+  dg_match.bins_[i_bin] = original_bin;
+}
+
+//! Helper function to retrieve fission q value from a nuclide
+
+double get_nuc_fission_q(const Nuclide& nuc, const Particle* p, int score_bin)
+{
+  if (score_bin == SCORE_FISS_Q_PROMPT) {
+    if (nuc.fission_q_prompt_) {
+      return (*nuc.fission_q_prompt_)(p->E_last_);
+    }
+  } else if (score_bin == SCORE_FISS_Q_RECOV) {
+    if (nuc.fission_q_recov_) {
+      return (*nuc.fission_q_recov_)(p->E_last_);
+    }
+  }
+  return 0.0;
+}
+
+//! Helper function to score fission energy
+//
+//! Pulled out to support both the fission_q scores and energy deposition
+//! score
+
+double score_fission_q(const Particle* p, int score_bin, const Tally& tally,
+  double flux, int i_nuclide, double atom_density)
+{
+  if (tally.estimator_ == ESTIMATOR_ANALOG) {
+    const Nuclide& nuc {*data::nuclides[p->event_nuclide_]};
+    if (settings::survival_biasing) {
+      // No fission events occur if survival biasing is on -- need to
+      // calculate fraction of absorptions that would have resulted in
+      // fission scaled by the Q-value
+      if (p->neutron_xs_[p->event_nuclide_].absorption > 0) {
+        return p->wgt_absorb_ * get_nuc_fission_q(nuc, p, score_bin)
+               * p->neutron_xs_[p->event_nuclide_].fission * flux
+               / p->neutron_xs_[p->event_nuclide_].absorption;
+      }
+    } else {
+      // Skip any non-absorption events
+      if (p->event_ == EVENT_SCATTER) return 0.0;
+      // All fission events will contribute, so again we can use particle's
+      // weight entering the collision as the estimate for the fission
+      // reaction rate
+      if (p->neutron_xs_[p->event_nuclide_].absorption > 0) {
+        return p->wgt_last_ * get_nuc_fission_q(nuc, p, score_bin)
+               * p->neutron_xs_[p->event_nuclide_].fission * flux
+               / p->neutron_xs_[p->event_nuclide_].absorption;
+      }
+    }
+  } else {
+    if (i_nuclide >= 0) {
+      const Nuclide& nuc {*data::nuclides[i_nuclide]};
+      return get_nuc_fission_q(nuc, p, score_bin) * atom_density * flux
+             * p->neutron_xs_[i_nuclide].fission;
+    } else {
```

Comment:
Optional: You can collapse this `else` block and the `if` block below into a single `else if` block.   E.g.,:


```
} else if (p-material_ != MATERIAL_VOID) {
  ...
  return score * flux;
}
return 0.0
```